Fix empty time durations

uklotzde opened the pull request:
Qt5: If QTime is null it stays null even if you add something to it.

When searching for "QTime()" I found some other minor issues that I fixed.

uklotzde committed:
Fix formatting of time durations

uklotzde committed:
Eliminate unnecessary heap allocation of QTime

uklotzde committed:
Fix missing pointer initialization

Pegasus-RPG commented:
You ran the tests with this change built against Qt4.x too, right?

uklotzde committed:
Delete obsolete dependency from TrackMetadata

uklotzde committed:
Move definition of Time::formatSeconds() into .cpp file

uklotzde commented:
Not yet. But I hope the build server does it for me ;) A local Qt5 build is 
still running.

I pushed some additional small fixes to clean up the code that uses 
time.h/.cpp. Now the plugin build does no longer depend on 
Time::formatSeconds() to be inlined.

On 01/08/2016 07:07 PM, Sean M. Pappalardo wrote:
>
> You ran the tests with this change built against Qt4.x too, right?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/845#issuecomment-170076652>.
>

rryan commented:
> Not yet. But I hope the build server does it for me ;) A local Qt5 build is
> still running.

Yea, Travis being green is good enough to tell us it's fine.

daschuer commented on src/util/time.cpp:
```diff
@@ -6,3 +6,33 @@ LLTIMER Time::s_timer;
+            (days > 0 ? (QString::number(days) %
+                         QLatin1String("'d', ")) : QString()) %
+            QLatin1String(days > 0 || t.hour() > 0 ? "hh:mm:ss" : "mm:ss") %
+            QLatin1String(showCentis ? ".zzz" : "");
```
Unrelated, but I have just noticed it: 
http://stackoverflow.com/questions/31477309/timestamp-in-iso-8601-the-last-6-digits-yyyy-mm-ddthhmmss
it should be QLocale::decimalPoint()

uklotzde replied:
QLocale::decimalPoint() is a separate issue.

But I added a unit test and it turns out that the formatting algorithm is 
wrong! Fix and tests will follow.

On 01/08/2016 09:47 PM, Daniel Schürmann wrote:
>
> In src/util/time.cpp 
> <https://github.com/mixxxdj/mixxx/pull/845#discussion_r49236294>:
>
> > +    if (dSeconds < 0) {
> > +        return "?";
> > +    }
> > +
> > +    const int days = static_cast<int>(dSeconds) / kSecondsPerDay;
> > +    dSeconds -= days * kSecondsPerDay;
> > +
> > +    // NOTE(uklotzde): Time() constructs a 'null' object, but
> > +    // we need 'zero' here.
> > +    QTime t = QTime(0, 0).addMSecs(dSeconds * kMillisPerSecond);
> > +
> > +    QString formatString =
> > +            (days > 0 ? (QString::number(days) %
> > +                         QLatin1String("'d', ")) : QString()) %
> > +            QLatin1String(days > 0 || t.hour() > 0 ? "hh:mm:ss" : "mm:ss") %
> > +            QLatin1String(showCentis ? ".zzz" : "");
>
> Unrelated, but I have just noticed it:
> http://stackoverflow.com/questions/31477309/timestamp-in-iso-8601-the-last-6-digits-yyyy-mm-ddthhmmss
> it should be QLocale::decimalPoint()
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/845/files#r49236294>.
>

daschuer commented:
LGTM.  If you think we should fix the decimal point issue later just give a hint.

rryan commented:
I would suggest fixing a separate issue in a separate PR :). Travis says :+1: so I say LGTM :)

uklotzde commented:
I've just added the formatting fix and the unit tests. Bug: The micro/centi 
seconds were missing in some cases.

Minor issue: I've also removed the strayed single quotes that where placed 
around days postfix "d".

On 01/08/2016 10:23 PM, RJ Ryan wrote:
>
> I would suggest fixing a separate issue in a separate PR :). Travis says 
> :+1: so I say LGTM :)
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/845#issuecomment-170127717>.
>

uklotzde commented:
You were right, RJ! We better fix those inconsistencies in a separate task. 
I've reverted my last commit immediately.

Strangely the micro/centi seconds are only added for large durations >= 1 
hour. But the library view looks strange with microsecond durations on 
single tracks. Looks like we need more formatting options.

On 01/08/2016 10:46 PM, Uwe Klotz wrote:
> I've just added the formatting fix and the unit tests. Bug: The 
> micro/centi seconds were missing in some cases.
>
> Minor issue: I've also removed the strayed single quotes that where placed 
> around days postfix "d".
>
> On 01/08/2016 10:23 PM, RJ Ryan wrote:
>>
>> I would suggest fixing a separate issue in a separate PR :). Travis says 
>> :+1: so I say LGTM :)
>>
>> —
>> Reply to this email directly or view it on GitHub 
>> <https://github.com/mixxxdj/mixxx/pull/845#issuecomment-170127717>.
>>
>

uklotzde commented:
Finished. The formatting fixes will follow in a separate PR.

daschuer commented:
Thank you very much. A nice bug, hard to spot.
LGTM!
